xmodule/services.py

```diff
@@ -308,3 +312,133 @@ def _handle_deprecated_progress_event(self, block, event):
+        return False
+
+    def answer_available(self):
+        """
+        Returns whether correctness is available now, for the given attributes.
+        """
+        if self._xblock:
+            show_answer = self._xblock.showanswer
+            max_attempts = self._xblock.max_attempts
+            attempts = self._xblock.attempts
+            is_correct = self._xblock.is_correct()
+            required_attempts = self._xblock.attempts_before_showanswer_button
+        else:
+            show_answer = ''
+            max_attempts = 0
+            attempts = 0
+            required_attempts = 0
+            is_correct = False
+        has_due_date = getattr(self._xblock, 'close_date', None) is not None
+        past_due = self.is_past_due()
+        is_attempted = self.is_attempted()
+        used_all_attempts = self.used_all_attempts()
+        closed = self.closed()
+
+        if not self.correctness_available():
+            # If correctness is being withheld, then don't show answers either.
+            return False
+        elif show_answer == ShowAnswer.NEVER:
+            return False
+        elif self._user_is_staff:
+            # This is after the 'never' check because course staff can see correctness
+            # unless the sequence/problem explicitly prevents it
+            return True
+        elif show_answer == ShowAnswer.ATTEMPTED:
+            return is_attempted or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.ANSWERED:
+            # NOTE: this is slightly different from 'attempted' -- resetting the problems
+            # makes lcp.done False, but leaves attempts unchanged.
+            return is_correct
+        elif show_answer == ShowAnswer.CLOSED:
+            return closed
+        elif show_answer == ShowAnswer.FINISHED:
+            return closed or is_correct
+        elif show_answer == ShowAnswer.CORRECT_OR_PAST_DUE:
+            return is_correct or (has_due_date and past_due)
+        elif show_answer == ShowAnswer.PAST_DUE:
+            return past_due
+        elif show_answer == ShowAnswer.AFTER_SOME_NUMBER_OF_ATTEMPTS:
+            if max_attempts and required_attempts >= max_attempts:
+                required_attempts = max_attempts
+            return attempts >= required_attempts
+        elif show_answer == ShowAnswer.ALWAYS:
+            return True
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS:
+            return used_all_attempts
+        elif show_answer == ShowAnswer.AFTER_ALL_ATTEMPTS_OR_CORRECT:
+            return used_all_attempts or is_correct
+        elif show_answer == ShowAnswer.ATTEMPTED_NO_PAST_DUE:
+            return is_attempted
+        return False
+
+    def correctness_available(self):
+        """
+        Returns whether correctness is available now, for the given attributes.
+        """
+        if self._xblock:
+            show_correctness = self._xblock.show_correctness
+        else:
+            show_correctness = ''
```

Comment:
Strictly speaking, isn't it the same as the original implementation? I.e. would always return `True` when `show_correctness` is `''`, and also we would be ignoring the `is_staff` check.

https://github.com/openedx/edx-platform/blob/794c6781fad8d290b548392a94f31a2d1b18e86f/xmodule/graders.py#L487-L503

I'm not against this suggestion. I was just trying to follow David's advice, i.e. to avoid changing the behavior, since it's a refactoring PR.